hatchways/team-long-john
# Feature create appointment integration #25 + FS: Refactor to have public accessible appointments #62

### What this PR does (required):
- Available time slots taking appointments in db and google into account.
- Creating and deleting events.
- Creating and deleting events affecting host user's google calendar.
- Creating and deleting google calendar events for the user.
- All of the aforementioned scheduler functionalities are publicly accessible.

### Screenshots / Videos (required):
- Scheduler for a specific meeting shared by the host user is publicly accessible through `/shared/:userName/:meetingId`
Example URL: `localhost:3000/shared/test/60f323381a376a60c971ca4a`
![image](https://user-images.githubusercontent.com/43277040/126046717-a8391a79-300f-4c8b-8d43-4d2036939c08.png)

- Available time slots takes consideration of host user's pre-existing appointments in the DB and google events.
![image](https://user-images.githubusercontent.com/43277040/126046760-c87ba613-eb0c-457b-8e77-6f212c704501.png)
![image](https://user-images.githubusercontent.com/43277040/126046767-e510688f-dd5d-4d9b-aeb4-7954f00fdb90.png)

-  Making a new appointment.
![image](https://user-images.githubusercontent.com/43277040/126046783-fcf502c7-4bd0-4eaa-91d2-e9807c6f52b7.png)

Without a valid email and (preferred) name, users can't make an appointment.
![image](https://user-images.githubusercontent.com/43277040/126046803-51e63f15-7bae-4a78-8968-6c8daf8fd313.png)

User can also change the timezone and make an appointment for convinience.
![image](https://user-images.githubusercontent.com/43277040/126047195-28930d76-ac3e-4244-825d-d0a9db5a4089.png)

- Appointment completion page. 
![image](https://user-images.githubusercontent.com/43277040/126046818-00693af4-ebb7-4203-9aa9-485ecc00615b.png)

Upon completion, corresponding appointment data is added to the DB. As it can be observed, appointment id is used for appointment completion page so it could be accessed again by the user in case the user wants to cancel or reschedule.
![image](https://user-images.githubusercontent.com/43277040/126046888-fa74f936-7cdf-4c6a-848f-173f2e54021f.png)
![image](https://user-images.githubusercontent.com/43277040/126046934-f157a916-975e-4429-a6eb-f738a7823f45.png)

Upon completion, a google event is created in host user's google calendar.
![image](https://user-images.githubusercontent.com/43277040/126046849-23df256f-9bdb-42e0-880f-d056ce7cbc57.png)

Even if different timezone was selected, on host user's perspective, scheduling will be done at the right time (July 23 3AM in Tokyo is July 22 2PM for the user's timezone - EDT). 
![image](https://user-images.githubusercontent.com/43277040/126047217-c5ad0edc-6ce7-42ed-9a07-36fd1f2da3ec.png)
![image](https://user-images.githubusercontent.com/43277040/126047226-ee979e82-a300-47cb-887f-f8d895d21fea.png)

- Google event could be created for the user if "Add to Google Calendar" button is clicked.
![image](https://user-images.githubusercontent.com/43277040/126046945-20cf5fe4-f070-48c0-99f6-7fbe14ceae80.png)

User's own google calendar
![image](https://user-images.githubusercontent.com/43277040/126046958-3e323265-a106-41d5-9d70-9ceb3c668444.png)

- Rescheduling or cancelling the event.
![image](https://user-images.githubusercontent.com/43277040/126046982-e453c424-cb6c-41f9-9656-978baa9a8927.png)

Rescheduling takes the user back to the same scheduler and cancelling takes the user back to the login.

Doing so also removes corresponding google event from the host user, and from the user (if the user added the event to own google calendar).
![image](https://user-images.githubusercontent.com/43277040/126047005-d11967ac-ef9d-4f35-9709-cf9828c4c234.png)
![image](https://user-images.githubusercontent.com/43277040/126047009-184a52af-e09c-4f50-8d88-d74ca13dc08c.png)

### Any information needed to test this feature (required):
- cd `/client` and run `npm install` (also, might need to run `npm update --force` if running into chokidar version issue). Then run `npm start`.
- cd `/server` and run `npm install`. Then run `npm run dev`.
- Go to `localhost:3000/shared/test/60f323381a376a60c971ca4a`.
- Make/delete an appointment. To check if the event was added/deleted on google calendar, sign into the test account pinned on slack group chat. You can also add the event to your own google account by clicking "Add to Google Calendar" button. 

### Any issues with the current functionality (optional):
- Sharing your own meeting is currently unavailable since dashboard integration is incomplete. You will be ale to share your own meeting after that task is complete.
- I tried to find and resolve all of the critical warning and bugs I could find. However, there may be more that I have missed. If you find one, please let me know.
